File: cap14/python/personaOculta.py
from datetime import datetime
from dateutil.relativedelta import relativedelta

class PersonaOculta:

    def __init__(self, nombre, apellidos, fechaNac, colorFavorito):
        self.__nombre = nombre
        self.__apellidos = apellidos
        self.__fechaNac = fechaNac
        self.__colorFavorito = colorFavorito
        print("---", nombre, "nació el", fechaNac) # debug

    def esAdulto(self):
        edad = self.__edad()
        return edad.years >= 18

    def cumpleHoy(self):
        edad = self.__edad()
        return edad.months == 0 and edad.days == 0

    def nombre(self):
        return self.__nombre

    @property
    def apellidos(self):
        return self.__apellidos

    @property
    def colorFavorito(self):
        return self.__colorFavorito

    @colorFavorito.setter
    def colorFavorito(self, nuevoColor):
        self.__colorFavorito = nuevoColor

    def __edad(self):
        return relativedelta(datetime.now(), self.__fechaNac)
### Fin class PersonaOculta

edadHoy17 = datetime.now() - relativedelta(years = 17)
edadHoy20YMedio = datetime.now() \
    - relativedelta(months = 6) - relativedelta(years = 20)

p1 = PersonaOculta("Paula", "Gala", edadHoy17, "rojo")
print(p1.nombre(), "es mayor de edad:", p1.esAdulto())
if p1.cumpleHoy():
    print("Hoy hay que felicitar a", p1.nombre(), p1.apellidos)

p2 = PersonaOculta("Pablo", "Prego", edadHoy20YMedio, "azul")
print(p2.nombre(), "es mayor de edad:", p2.esAdulto())
if p2.cumpleHoy():
    print("Hoy hay que felicitar a", p2.nombre, p2.apellidos)

print("Colores Favoritos: " , p1.nombre(), p1.colorFavorito, 
      p2.nombre(), p2.colorFavorito)
# Ahora a Pablo le gusta el color favorito de Paula
p2.colorFavorito = p1.colorFavorito
print("Colores Favoritos: " , p1.nombre(), p1.colorFavorito, 
      p2.nombre(), p2.colorFavorito)

File: cap14/java/oo/PersonaOculta.java
package oo;

import java.time.LocalDate;
import java.time.Period;

public class PersonaOculta {
	private String nombre;
	private String apellidos;
	private LocalDate fechaNac;
	private String colorFavorito;
	
	public PersonaOculta(String nombre, String apellidos, 
			LocalDate fechaNac, String colorFavorito) {
		this.nombre = nombre;
		this.apellidos = apellidos;
		this.fechaNac = fechaNac;
		this.colorFavorito = colorFavorito;
		System.out.println("--- " + nombre + " " + apellidos 
				+ " nació el " + fechaNac);
	}

	public boolean esAdulto() {
		return edad().getYears() >= 18;
	}
	
	public boolean cumpleHoy() {
		return edad().getMonths() == 0 && edad().getDays() == 0;
	}
	
	public String getNombre() {
		return nombre;
	}
	
	public String getApellidos() {
		return apellidos;
	}
	
	public String getColorFavorito() {
		return colorFavorito;
	}
	
	public void setColorFavorito(String colorFavorito) {
		this.colorFavorito = colorFavorito;
	}
	
	private Period edad() {
		return Period.between(fechaNac, LocalDate.now());
	}
}